File: core/parser/statement.ts
import { Node, NodeParser } from "./node.ts";
import { WhistleParser } from "./parser.ts";
import { Expression, ParseExpression } from "./expression.ts";
import { ParseTip, Tip } from "./tip.ts";

export type Statement =
  | IfStatement
  | LoopStatement
  | WhileStatement
  | ReturnStatement
  | ContinueStatement
  | BreakStatement
  | VariableDeclaration
  | ValueDeclaration
  | BlockStatement
  | ExpressionStatement
  | Tip;

export const ParseStatement: NodeParser<Statement> = (parser:
  WhistleParser) => {
  switch (parser.current.type) {
    case "Tip":
      return ParseTip(parser);
    case "Keyword":
      switch (parser.current.value) {
        case "if":
          return ParseIfStatement(parser);
        case "return":
          return ParseReturnStatement(parser);
        case "var":
          return ParseVariableDeclaration(parser);
        case "val":
          return ParseValueDeclaration(parser);
        case "loop":
          return ParseLoopStatement(parser);
        case "while":
          return ParseWhileStatement(parser);
        case "continue":
          return ParseContinueStatement(parser);
        case "break":
          return ParseBreakStatement(parser);
      }
    case "LeftBrace":
      return ParseBlockStatement(parser);
    default:
      return ParseExpressionStatement(parser);
  }

  throw `Could not parse statement ${JSON.stringify(parser.current)}`;
};

export interface ReturnStatement extends Node<Expression> {
  type: "ReturnStatement";
}

export const ParseReturnStatement: NodeParser<ReturnStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "return" });

  return {
    type: "ReturnStatement",
    value: ParseExpression(parser),
  };
};

export interface ContinueStatement extends Node<undefined> {
  type: "ContinueStatement";
}

export const ParseContinueStatement: NodeParser<ContinueStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "continue" });

  return {
    type: "ContinueStatement",
    value: undefined,
  };
};

export interface BreakStatement extends Node<undefined> {
  type: "BreakStatement";
}

export const ParseBreakStatement: NodeParser<BreakStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "break" });

  return {
    type: "BreakStatement",
    value: undefined,
  };
};

export interface IfStatement extends Node<{
  condition: Expression;
  then: Statement;
  else?: Statement;
}> {
  type: "IfStatement";
}

export const ParseIfStatement: NodeParser<IfStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "if" });

  return {
    type: "IfStatement",
    value: {
      condition: ParseExpression(parser),
      then: ParseStatement(parser),
      else: parser.is({ type: "Keyword", value: "else" })
        ? parser.eat({ type: "Keyword", value: "else" }) &&
          ParseStatement(parser)
        : undefined,
    },
  };
};

export interface LoopStatement extends Node<Statement> {
  type: "LoopStatement";
}

export const ParseLoopStatement: NodeParser<LoopStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "loop" });

  return {
    type: "LoopStatement",
    value: ParseStatement(parser),
  };
};

export interface WhileStatement extends Node<{
  condition: Expression;
  then: Statement;
}> {
  type: "WhileStatement";
}

export const ParseWhileStatement: NodeParser<WhileStatement> = (parser:
  WhistleParser) => {
  parser.eat({ type: "Keyword", value: "while" });

  return {
    type: "WhileStatement",
    value: {
      condition: ParseExpression(parser),
      then: ParseStatement(parser),
    },
  };
};

export interface VariableDeclaration extends Node<{
  name: string;
  type: string;
  value: Expression;
}> {
  type: "VariableDeclaration";
}

export const ParseVariableDeclaration: NodeParser<VariableDeclaration> = (
  parser: WhistleParser,
) => {
  parser.eat({ type: "Keyword", value: "var" });

  const name = parser.eat({ type: "Identifier" }).value;

  parser.eat({ type: "Colon" });

  const type = parser.eat({ type: "Type" }).value;

  parser.eat({ type: "Operator", value: "=" });

  const value = ParseExpression(parser);

  return {
    type: "VariableDeclaration",
    value: {
      name,
      type,
      value,
    },
  };
};

export interface ValueDeclaration extends Node<{
  name: string;
  type: string;
  value: Expression;
}> {
  type: "ValueDeclaration";
}

export const ParseValueDeclaration: NodeParser<ValueDeclaration> = (
  parser: WhistleParser,
) => {
  parser.eat({ type: "Keyword", value: "val" });

  const name = parser.eat({ type: "Identifier" }).value;

  parser.eat({ type: "Colon" });

  const type = parser.eat({ type: "Type" }).value;

  parser.eat({ type: "Operator", value: "=" });

  const value = ParseExpression(parser);

  return {
    type: "ValueDeclaration",
    value: {
      name,
      type,
      value,
    },
  };
};

export interface BlockStatement extends Node<Statement[]> {
  type: "BlockStatement";
}

export const ParseBlockStatement: NodeParser<BlockStatement> = (parser:
  WhistleParser) => {
  const statements: Statement[] = [];

  parser.eat({ type: "LeftBrace" });

  while (!parser.is({ type: "RightBrace" })) {
    statements.push(ParseStatement(parser));
  }

  parser.eat({ type: "RightBrace" });

  return {
    type: "BlockStatement",
    value: statements,
  };
};

export interface ExpressionStatement extends Node<Expression> {
  type: "ExpressionStatement";
}

export const ParseExpressionStatement: NodeParser<ExpressionStatement> =
  (parser: WhistleParser) => {
    return {
      type: "ExpressionStatement",
      value: ParseExpression(parser),
    };
  };

File: core/src/parser/program/statement.rs
use crate::parser::ast::*;
use crate::parser::Parser;
use super::expression::*;
use super::ident::*;
use crate::lexer::*;

pub fn parse_statement(parser: &mut Parser) -> Option<Stmt> {
    //TODO: parse_ident
    match parser.peek() {
        Some(Token::Tip(_tip)) => parse_tip(parser),
        Some(Token::Keyword(keyword)) => {
            match keyword {
                Keyword::If => parse_if_statement(parser),
                Keyword::Return => parse_return_statement(parser),
                Keyword::Var => parse_var_statement(parser),
                Keyword::Val => parse_val_statement(parser),
                Keyword::While => parse_while_statement(parser),
                Keyword::Continue => parse_continue_statement(parser),
                Keyword::Break => parse_break_statement(parser),
                _ => {println!("Could not parse statement {:?}", keyword); None}
            }
        }
        Some(Token::Punc(Punc::LeftBrace)) => parse_block_statement(parser),
        _ => {println!("Could not parse statement {:?}", parser.peek()); None}
      }
}

//TODO: parse_tip
pub fn parse_tip(parser:&mut Parser) -> Option<Stmt> {
    Some(Stmt::Continue)
}

pub fn parse_if_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::If)).is_some() {
        if let Some(cond) = parse_expression(parser) {
            if let Some(then_stmt) = parse_statement(parser) {
                return Some(Stmt::If {
                    cond: Box::new(cond),
                    then_stmt: Box::new(then_stmt),
                    else_stmt: parse_else_statement(parser)
                })
            }
        }
    }
    None
}

pub fn parse_else_statement(parser:&mut Parser) -> Option<Box<Stmt>> {
    if parser.eat_tok(Token::Keyword(Keyword::Else)).is_some() {
        if let Some(else_stmt) = parse_statement(parser) {
            return Some(Box::new(else_stmt))
        }
    }
    None
}

pub fn parse_return_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::Return)).is_some() {
        if let Some(expr) = parse_expression(parser) {
            return Some(Stmt::Return(Some(Box::new(expr))))
        }
    }
    None
}

pub fn parse_var_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::Var)).is_some() {
        if let Some(ident_typed) = parse_ident_typed(parser) {
            if let Some(value) = parse_assign(parser) {
                return Some(Stmt::VarDecl {
                    ident_typed: ident_typed,
                    val: Box::new(value)
                })
            }
        }
    }
    None
}

pub fn parse_assign(parser:&mut Parser) -> Option<Expr> {
    if parser.eat_tok(Token::Operator(Operator::Assign)).is_some() {
        return parse_expression(parser)
    }
    None
}

pub fn parse_val_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::Val)).is_some() {
        if let Some(ident_typed) = parse_ident_typed(parser) {
            if let Some(value) = parse_assign(parser) {
                return Some(Stmt::ValDecl {
                    ident_typed: ident_typed,
                    val: Box::new(value)
                })
            }
        }
    }
    None
}

pub fn parse_while_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::If)).is_some() {
        if let Some(cond) = parse_expression(parser) {
            if let Some(do_stmt) = parse_statement(parser) {
                return Some(Stmt::While {
                    cond: Some(Box::new(cond)),
                    do_stmt: Box::new(do_stmt),
                })
            }
        }
    }
    None
}

pub fn parse_continue_statement(parser:&mut Parser) -> Option<Stmt> {
    parser.eat_tok(Token::Keyword(Keyword::Continue));
    return Some(Stmt::Continue)
}

pub fn parse_break_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Keyword(Keyword::Break)).is_some() {
        return Some(Stmt::Break)
    }
    None
}

pub fn parse_block_statement(parser:&mut Parser) -> Option<Stmt> {
    if parser.eat_tok(Token::Punc(Punc::LeftBrace)).is_some() {
        let statements = parser.until_is(parse_statement, Token::Punc(Punc::RightBrace));
        if parser.eat_tok(Token::Punc(Punc::RightBrace)).is_some() {
            return Some(Stmt::Block(statements))
        }
    }
    None
}

pub fn parse_expression_statement(parser:&mut Parser) -> Option<Stmt> {
    if let Some(expr) = parse_expression(parser) {
        return Some(Stmt::Expr(expr))
    }
    None
}